File: apps/browser-extension/src/crypto/key-envelope.ts
import { Chacha20Poly1305 } from "@hpke/chacha20poly1305";
import { CipherSuite, HkdfSha256 } from "@hpke/core";
import { DhkemX25519HkdfSha256 } from "@hpke/dhkem-x25519";

import {
  keyEpochId as deriveKeyEpochId,
  type Identifier,
  identifier,
  keyEnvelopeId,
} from "../domain/canonical/identifiers";
import {
  byteString,
  exactCode,
  exactMap,
  identifierValue,
  mapValue,
  nonnegativeInteger,
  nullable,
  oneOfCodes,
} from "../domain/canonical/schema";
import { concatBytes, transcript } from "../domain/canonical/transcript";
import {
  canonicalMap,
  decodeCanonicalValue,
  encodeCanonicalValue,
} from "../domain/canonical/value";
import { bytesEqual } from "../domain/hash";
import {
  COMPACT_STORAGE_CLASS,
  decodeOpaqueEnvelope,
  encodeOpaqueEnvelope,
  type OpaqueEnvelope,
} from "../storage/opaque-envelope";
import { CryptoOperationError } from "./errors";

export const KEY_ENVELOPE_FORMAT = 1 as const;
export type KeyEnvelopeTargetKind = 1 | 2;

export interface KeyEnvelopePlaintext {
  readonly vaultId: Identifier<"Vault">;
  readonly keyEpochId: Identifier<"KeyEpoch">;
  readonly keyEpochKey: Uint8Array;
  readonly targetKind: KeyEnvelopeTargetKind;
  readonly targetCredentialId: Identifier<"RecoveryCredential" | "ClientCredential">;
  readonly targetRevision: number | null;
  readonly bytes: Uint8Array;
  readonly id: Identifier<"KeyEnvelope">;
}

export interface SealedKeyEnvelope extends KeyEnvelopePlaintext {
  readonly envelope: OpaqueEnvelope;
}

const suite = new CipherSuite({
  kem: new DhkemX25519HkdfSha256(),
  kdf: new HkdfSha256(),
  aead: new Chacha20Poly1305(),
});

function hpkeInfo(targetKind: KeyEnvelopeTargetKind, padding: Uint8Array): Uint8Array {
  return transcript(
    targetKind === 1 ? "awsm:recovery-key-envelope-hpke:v1" : "awsm:client-key-envelope-hpke:v1",
    [padding],
  );
}

function validateTarget(targetKind: KeyEnvelopeTargetKind, targetRevision: number | null): void {
  if ((targetKind === 1) !== (targetRevision !== null)) {
    throw new TypeError("Key Envelope revision must exist only for a Recovery Credential");
  }
}

function encodePlaintext(input: {
  readonly vaultId: Identifier<"Vault">;
  readonly keyEpochId: Identifier<"KeyEpoch">;
  readonly keyEpochKey: Uint8Array;
  readonly targetKind: KeyEnvelopeTargetKind;
  readonly targetCredentialId: Identifier<"RecoveryCredential" | "ClientCredential">;
  readonly targetRevision: number | null;
}): Uint8Array {
  validateTarget(input.targetKind, input.targetRevision);
  if (!bytesEqual(deriveKeyEpochId(input.vaultId, input.keyEpochKey), input.keyEpochId)) {
    throw new TypeError("Key Envelope Key Epoch ID does not match its key and Vault");
  }
  const expectedKind = input.targetKind === 1 ? "RecoveryCredential" : "ClientCredential";
  identifier(expectedKind, input.targetCredentialId);
  if (
    input.targetRevision !== null &&
    (!Number.isSafeInteger(input.targetRevision) || input.targetRevision < 0)
  ) {
    throw new TypeError("Recovery Credential revision must be nonnegative");
  }
  return encodeCanonicalValue(
    canonicalMap([
      [0, KEY_ENVELOPE_FORMAT],
      [1, input.vaultId],
      [2, input.keyEpochId],
      [3, input.keyEpochKey],
      [4, input.targetKind],
      [5, input.targetCredentialId],
      [6, input.targetRevision],
    ]),
  );
}

function decodePlaintext(bytes: Uint8Array): KeyEnvelopePlaintext {
  const map = exactMap(
    decodeCanonicalValue(bytes),
    [0, 1, 2, 3, 4, 5, 6],
    "Key Envelope plaintext",
  );
  exactCode(mapValue(map, 0), KEY_ENVELOPE_FORMAT, "Key Envelope format");
  const vaultId = identifierValue(mapValue(map, 1), "Vault", "Key Envelope Vault ID");
  const epochId = identifierValue(mapValue(map, 2), "KeyEpoch", "Key Envelope Epoch ID");
  const epochKey = byteString(mapValue(map, 3), 32, "Key Envelope Epoch Key");
  if (!bytesEqual(deriveKeyEpochId(vaultId, epochKey), epochId)) {
    throw new TypeError("Opened Key Envelope has an invalid Key Epoch binding");
  }
  const targetKind = oneOfCodes(mapValue(map, 4), [1, 2] as const, "Key Envelope target kind");
  const targetCredentialId = identifierValue(
    mapValue(map, 5),
    targetKind === 1 ? "RecoveryCredential" : "ClientCredential",
    "Key Envelope target Credential ID",
  );
  const targetRevision = nullable(mapValue(map, 6), (revision) =>
    nonnegativeInteger(revision, "Key Envelope target revision"),
  );
  validateTarget(targetKind, targetRevision);
  const canonicalBytes = Uint8Array.from(bytes);
  return {
    vaultId,
    keyEpochId: epochId,
    keyEpochKey: epochKey,
    targetKind,
    targetCredentialId,
    targetRevision,
    bytes: canonicalBytes,
    id: keyEnvelopeId(canonicalBytes),
  };
}

export async function sealKeyEnvelope(input: {
  readonly vaultId: Identifier<"Vault">;
  readonly keyEpochId: Identifier<"KeyEpoch">;
  readonly keyEpochKey: Uint8Array;
  readonly targetKind: KeyEnvelopeTargetKind;
  readonly targetCredentialId: Identifier<"RecoveryCredential" | "ClientCredential">;
  readonly targetRevision: number | null;
  readonly recipientWrappingPublicKey: Uint8Array;
  readonly outerPadding?: Uint8Array;
}): Promise<SealedKeyEnvelope> {
  const plaintextBytes = encodePlaintext(input);
  const plaintext = decodePlaintext(plaintextBytes);
  const padding = input.outerPadding
    ? byteString(input.outerPadding, 32, "Key Envelope outer padding")
    : crypto.getRandomValues(new Uint8Array(32));
  const publicKey = await suite.kem.deserializePublicKey(
    byteString(input.recipientWrappingPublicKey, 32, "Recipient wrapping public key"),
  );
  let sealed: { readonly enc: ArrayBuffer; readonly ct: ArrayBuffer };
  try {
    sealed = await suite.seal(
      { recipientPublicKey: publicKey, info: hpkeInfo(input.targetKind, padding) },
      plaintextBytes,
      new Uint8Array(),
    );
  } catch {
    throw new CryptoOperationError();
  }
  const encapsulatedKey = new Uint8Array(sealed.enc);
  if (encapsulatedKey.byteLength !== 32) throw new CryptoOperationError();
  const envelope = encodeOpaqueEnvelope({
    storageClass: COMPACT_STORAGE_CLASS,
    protectionParameters: concatBytes([encapsulatedKey, padding]),
    payload: new Uint8Array(sealed.ct),
  });
  return { ...plaintext, envelope };
}

export async function openKeyEnvelope(input: {
  readonly targetKind: KeyEnvelopeTargetKind;
  readonly recipientWrappingPrivateKey: Uint8Array;
  readonly envelopeBytes: Uint8Array;
}): Promise<KeyEnvelopePlaintext & { readonly envelope: OpaqueEnvelope }> {
  const envelope = decodeOpaqueEnvelope(input.envelopeBytes);
  if (envelope.storageClass !== COMPACT_STORAGE_CLASS) {
    throw new TypeError("Key Envelope must use Compact storage");
  }
  const encapsulatedKey = envelope.protectionParameters.slice(0, 32);
  const padding = envelope.protectionParameters.slice(32);
  const privateKey = await suite.kem.deserializePrivateKey(
    byteString(input.recipientWrappingPrivateKey, 32, "Recipient wrapping private key"),
  );
  let plaintextBytes: Uint8Array;
  try {
    plaintextBytes = new Uint8Array(
      await suite.open(
        {
          recipientKey: privateKey,
          enc: encapsulatedKey,
          info: hpkeInfo(input.targetKind, padding),
        },
        envelope.payload,
        new Uint8Array(),
      ),
    );
  } catch {
    throw new CryptoOperationError();
  }
  const plaintext = decodePlaintext(plaintextBytes);
  if (plaintext.targetKind !== input.targetKind) throw new CryptoOperationError();
  return { ...plaintext, envelope };
}

File: apps/runtime-go/internal/crypto/key_envelope.go
package awsmcrypto

import (
	cryptorand "crypto/rand"
	"crypto/sha256"
	"errors"
	"fmt"
	"io"

	"github.com/mashuproject/awsm_bak/apps/runtime-go/internal/canonical"
	"github.com/mashuproject/awsm_bak/apps/runtime-go/internal/storage"
)

const (
	RecoveryCredentialTarget uint64 = 1
	ClientCredentialTarget   uint64 = 2
)

type KeyEnvelopeInput struct {
	VaultID                    [32]byte
	KeyEpochID                 [32]byte
	KeyEpochKey                []byte
	TargetKind                 uint64
	TargetCredentialID         [32]byte
	TargetRevision             *uint64
	RecipientWrappingPublicKey []byte
	Padding                    []byte
	EphemeralSeed              []byte
}

type KeyEnvelope struct {
	VaultID              [32]byte
	KeyEpochID           [32]byte
	KeyEpochKey          []byte
	TargetKind           uint64
	TargetCredentialID   [32]byte
	TargetRevision       *uint64
	ProtectionParameters []byte
	Bytes                []byte
	ID                   [32]byte
	Envelope             storage.OpaqueEnvelope
}

func SealKeyEnvelope(input KeyEnvelopeInput) (KeyEnvelope, error) {
	if err := validateKeyEnvelopeTarget(input.TargetKind, input.TargetRevision); err != nil {
		return KeyEnvelope{}, err
	}
	if input.TargetCredentialID == ([32]byte{}) {
		return KeyEnvelope{}, errors.New("Key Envelope target Credential ID must not be zero")
	}
	if len(input.KeyEpochKey) != 32 {
		return KeyEnvelope{}, errors.New("Key Envelope Key Epoch Key must contain exactly 32 bytes")
	}
	expectedEpoch, err := KeyEpochID(input.VaultID, input.KeyEpochKey)
	if err != nil {
		return KeyEnvelope{}, err
	}
	if expectedEpoch != input.KeyEpochID {
		return KeyEnvelope{}, errors.New("Key Envelope Key Epoch ID does not match its key and Vault")
	}
	plaintext, err := encodeKeyEnvelopePlaintext(input)
	if err != nil {
		return KeyEnvelope{}, err
	}
	padding := append([]byte(nil), input.Padding...)
	if len(padding) == 0 {
		padding = make([]byte, 32)
		if _, err := io.ReadFull(cryptorand.Reader, padding); err != nil {
			return KeyEnvelope{}, fmt.Errorf("generate Key Envelope padding: %w", err)
		}
	}
	if len(padding) != 32 {
		return KeyEnvelope{}, errors.New("Key Envelope padding must contain exactly 32 bytes")
	}
	info, err := keyEnvelopeInfo(input.TargetKind, padding)
	if err != nil {
		return KeyEnvelope{}, err
	}
	enc, ciphertext, err := HPKESeal(input.RecipientWrappingPublicKey, info, plaintext, nil, input.EphemeralSeed)
	if err != nil {
		return KeyEnvelope{}, err
	}
	protection := append(append([]byte(nil), enc...), padding...)
	outer, err := storage.EncodeOpaqueEnvelope(storage.OpaqueEnvelopeInput{
		StorageClass:         storage.CompactStorageClass,
		ProtectionParameters: protection,
		Payload:              ciphertext,
	})
	if err != nil {
		return KeyEnvelope{}, err
	}
	decoded, err := storage.DecodeOpaqueEnvelope(outer)
	if err != nil {
		return KeyEnvelope{}, err
	}
	return KeyEnvelope{
		VaultID: input.VaultID, KeyEpochID: input.KeyEpochID,
		KeyEpochKey: append([]byte(nil), input.KeyEpochKey...), TargetKind: input.TargetKind,
		TargetCredentialID: input.TargetCredentialID, TargetRevision: cloneRevision(input.TargetRevision),
		ProtectionParameters: protection, Bytes: append([]byte(nil), plaintext...),
		ID: sha256.Sum256(mustKeyEnvelopeTranscript(plaintext)), Envelope: decoded,
	}, nil
}

func OpenKeyEnvelope(targetKind uint64, recipientWrappingPrivateKey, envelopeBytes []byte) (KeyEnvelope, error) {
	if targetKind != RecoveryCredentialTarget && targetKind != ClientCredentialTarget {
		return KeyEnvelope{}, errors.New("Key Envelope target kind is unknown")
	}
	envelope, err := storage.DecodeOpaqueEnvelope(envelopeBytes)
	if err != nil {
		return KeyEnvelope{}, err
	}
	if envelope.StorageClass != storage.CompactStorageClass || len(envelope.ProtectionParameters) != 64 {
		return KeyEnvelope{}, errors.New("Key Envelope outer representation is invalid")
	}
	padding := envelope.ProtectionParameters[32:]
	info, err := keyEnvelopeInfo(targetKind, padding)
	if err != nil {
		return KeyEnvelope{}, err
	}
	plaintext, err := HPKEOpen(recipientWrappingPrivateKey, envelope.ProtectionParameters[:32], info, envelope.Payload, nil)
	if err != nil {
		return KeyEnvelope{}, err
	}
	input, err := decodeKeyEnvelopePlaintext(plaintext)
	if err != nil {
		return KeyEnvelope{}, err
	}
	if input.TargetKind != targetKind {
		return KeyEnvelope{}, errors.New("opened Key Envelope target kind does not match the attempted context")
	}
	return KeyEnvelope{
		VaultID: input.VaultID, KeyEpochID: input.KeyEpochID,
		KeyEpochKey: append([]byte(nil), input.KeyEpochKey...), TargetKind: input.TargetKind,
		TargetCredentialID: input.TargetCredentialID, TargetRevision: cloneRevision(input.TargetRevision),
		ProtectionParameters: append([]byte(nil), envelope.ProtectionParameters...), Bytes: append([]byte(nil), plaintext...),
		ID: sha256.Sum256(mustKeyEnvelopeTranscript(plaintext)), Envelope: envelope,
	}, nil
}

func encodeKeyEnvelopePlaintext(input KeyEnvelopeInput) ([]byte, error) {
	var revision canonical.Value
	if input.TargetRevision != nil {
		revision = *input.TargetRevision
	}
	return canonical.EncodeValue(canonical.Map{
		0: uint64(1),
		1: append([]byte(nil), input.VaultID[:]...),
		2: append([]byte(nil), input.KeyEpochID[:]...),
		3: append([]byte(nil), input.KeyEpochKey...),
		4: input.TargetKind,
		5: append([]byte(nil), input.TargetCredentialID[:]...),
		6: revision,
	})
}

func decodeKeyEnvelopePlaintext(encoded []byte) (KeyEnvelopeInput, error) {
	value, err := canonical.DecodeValue(encoded)
	if err != nil {
		return KeyEnvelopeInput{}, err
	}
	if !compactIsMap(value) || len(compactMapKeys(value)) != 7 {
		return KeyEnvelopeInput{}, errors.New("Key Envelope plaintext contains missing or unknown fields")
	}
	var input KeyEnvelopeInput
	if input.VaultID, err = keyEnvelopeIdentifier(value, 1, "Vault ID"); err != nil {
		return KeyEnvelopeInput{}, err
	}
	if input.KeyEpochID, err = keyEnvelopeIdentifier(value, 2, "Key Epoch ID"); err != nil {
		return KeyEnvelopeInput{}, err
	}
	if input.KeyEpochKey, err = keyEnvelopeBytes(value, 3, "Key Epoch Key", 32); err != nil {
		return KeyEnvelopeInput{}, err
	}
	if input.TargetKind, err = keyEnvelopeNumber(value, 4, "target kind"); err != nil {
		return KeyEnvelopeInput{}, err
	}
	if input.TargetCredentialID, err = keyEnvelopeIdentifier(value, 5, "target Credential ID"); err != nil {
		return KeyEnvelopeInput{}, err
	}
	revisionValue, ok := compactMapValue(value, 6)
	if !ok {
		return KeyEnvelopeInput{}, errors.New("Key Envelope target revision is missing")
	}
	if revisionValue != nil {
		revision, ok := revisionValue.(uint64)
		if !ok {
			return KeyEnvelopeInput{}, errors.New("Key Envelope target revision is invalid")
		}
		input.TargetRevision = &revision
	}
	if err := validateKeyEnvelopeTarget(input.TargetKind, input.TargetRevision); err != nil {
		return KeyEnvelopeInput{}, err
	}
	expectedEpoch, err := KeyEpochID(input.VaultID, input.KeyEpochKey)
	if err != nil || expectedEpoch != input.KeyEpochID {
		return KeyEnvelopeInput{}, errors.New("Key Envelope Key Epoch binding is invalid")
	}
	return input, nil
}

func validateKeyEnvelopeTarget(kind uint64, revision *uint64) error {
	if kind != RecoveryCredentialTarget && kind != ClientCredentialTarget {
		return errors.New("Key Envelope target kind is unknown")
	}
	if (kind == RecoveryCredentialTarget) != (revision != nil) {
		return errors.New("Key Envelope revision must exist only for a Recovery Credential")
	}
	return nil
}

func keyEnvelopeInfo(kind uint64, padding []byte) ([]byte, error) {
	label := "awsm:client-key-envelope-hpke:v1"
	if kind == RecoveryCredentialTarget {
		label = "awsm:recovery-key-envelope-hpke:v1"
	}
	return canonical.Transcript(label, padding)
}

func keyEnvelopeIdentifier(value canonical.Value, key uint64, field string) ([32]byte, error) {
	bytesValue, err := keyEnvelopeBytes(value, key, field, 32)
	if err != nil {
		return [32]byte{}, err
	}
	var result [32]byte
	copy(result[:], bytesValue)
	if result == ([32]byte{}) {
		return [32]byte{}, fmt.Errorf("%s must not be zero", field)
	}
	return result, nil
}

func keyEnvelopeBytes(value canonical.Value, key uint64, field string, length int) ([]byte, error) {
	bytesValue, ok := compactMapBytes(value, key)
	if !ok || len(bytesValue) != length {
		return nil, fmt.Errorf("%s must contain exactly %d bytes", field, length)
	}
	return append([]byte(nil), bytesValue...), nil
}

func keyEnvelopeNumber(value canonical.Value, key uint64, field string) (uint64, error) {
	numeric, ok := compactMapNumber(value, key)
	if !ok {
		return 0, fmt.Errorf("%s must be an unsigned integer", field)
	}
	return numeric, nil
}

func cloneRevision(value *uint64) *uint64 {
	if value == nil {
		return nil
	}
	copyValue := *value
	return &copyValue
}

func mustKeyEnvelopeTranscript(plaintext []byte) []byte {
	value, err := canonical.Transcript("awsm:key-envelope-id:v1", plaintext)
	if err != nil {
		panic(err)
	}
	return value
}
